integration_tests/src/main/python/fastparquet_utils.py

```diff
@@ -0,0 +1,112 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import pytest
+
+from pyspark.sql import Row
+
+def get_fastparquet_result_converter():
+    # get a converter which converts fastparquet result to Spark cpu type result
+    # the convert is used to convet flatted Row to nested Row
+    # e.g.:
+    # Row(a.b1 = 1, a.b2 = 2) => Row(a = Row(b1 = 1, b2 = 2))
+    return _convert_fastparquet_result
+
+def _convert_fastparquet_result(fastparquet_cpu, gpu):
+    """
+    The result of fastparquet is different from the result of GPU, e.g.:
+        --- CPU OUTPUT
+        +++ GPU OUTPUT
+        @@ -1 +1 @@
+        -Row(a.first=-341142443, a.second=3.333994866005594e-37)
+        +Row(a=Row(first=-341142443, second=3.333994866005594e-37))
+    This method is used to canonicalize the fastparquet result
+    :param fastparquet_cpu: the fastparquet result, it's a Spark Row list,
+    :param gpu: the gpu result, it's a Spark Row list
+    :return: (converted cpu, gpu) tuple
+    """
+    assert(len(fastparquet_cpu) == len(gpu))
+    new_cpu = convert_fastparquet_result(fastparquet_cpu, gpu)
+    return (new_cpu, gpu)
+
+def convert_fastparquet_result(cpu, gpu):
+    return [_convert_fastparquet_row(cpu_row, gpu_row) for (cpu_row, gpu_row) in zip(cpu, gpu)]
+
+def _convert_fastparquet_row(cpu_row, gpu_row):
+    cpu_dict = cpu_row.asDict()
+    _merge_cpu_dict(cpu_dict, gpu_row.asDict())
+    return Row(**cpu_dict)
+
+def _merge_cpu_dict(cpu_dict, gpu_dict, cpu_prefix=""):
```

Comment:
Hmm, is ```gpu_dict``` a must parameter to pass in? It seems we can directly build from cpu_dict. It's more like a flatted representation of GPU format.

Reply:
gpu hierarchy is the baseline in this method, we try to build a hierarchy according to gpu hierarchy, so passed in `gpu_dict`.

Consider this case:
cpu: Row(a.first = 1, a.second = 2)
gpu: Row(a.first = 1, a.second = 2)
If only pass in cpu without gpu, then converted cpu will be Row(a = Row(first=1, second=2)), In this case gpu actually has column name "a.first", "a.second"

Reply:
NIT: It is possible to flatten the GPU result instead of reconstruct the CPU rows ? I guess flattening may be simpler than reconstruct the nested rows, and we just only need to compare the converted results.

Reply:
In this PR we convert the result to follow the result of Spark. The GPU result also follow the result type of Spark.
I think it's better to take the Spark result as the baseline.

`test_fastparquet` you provided should return a `data frame`,  `assert_gpu_and_cpu_are_equal_collect` expects the function retuns a `df` because `assert_gpu_and_cpu_are_equal_collect` will call `test_fastparquet().collect`.

Reply:
you are right.

Reply:
> d cpu will be Row(a = Row(first=1, second=2)), In this case gpu actually has column name "a.first", "a.second"

This is a case where we can't convert a flatten value back. Sounds like we only care about data value other than the "schema". Then, why not just comparing a canonicalized value directly without rebuilding the tree again?

Reply:
Yes, flatten the GPU result is simple and doable.
Since fastparquet already flattened the result and the schema is missed, the GPU result also can be convert to flatted format.